### PR TITLE
Wrote SparkMax wrapper

### PR DESCRIPTION
**Pull request by thehappycamden**
YORT

### CODE REVIEW

File: src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+| So go ahead and check out the video            |
+| description to find out more about             |
+| RAID: Shadow Legends™. There, you will find a  |
+| link to the store page and a special code to   |
+| unlock all sorts of goodies. Using the special |
+| code, you can get 50,000 Silver immediately,   |
+| and a FREE Epic Level Champion as part of the  |
+| new players program, courtesy of course of the |
+| RAID: Shadow Legends™ devs.                    |
+\************************************************/
```
Comment (pietroglyph) [inline src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java:32]: Remove this comment :)

File: src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+    @Override
+    public SpartronicsEncoder getEncoder() {
+        return mEncoder;
+    }
+
+    @Override
+    public double getVoltageOutput() {
+        return mSparkMax.getBusVoltage() * mSparkMax.get();
```
Comment (pietroglyph) [inline src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java]: Change to from `mSparkMax.get()` to `mSparkMax.getAppliedOutput()`.

File: src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        mUseMotionProfileForPosition = useMotionProfile;
+    }
+
+    @Override
+    public void setDutyCycle(double dutyCycle, double arbitraryFeedForwardVolts) {
+        if (mLastControlMode != ControlType.kDutyCycle) {
+            mLastControlMode = ControlType.kDutyCycle;
+        }
+        mSparkMax.getPIDController().setReference(dutyCycle, ControlType.kDutyCycle, 0, 
+        arbitraryFeedForwardVolts / mVoltageCompSaturation);
```
Comment (pietroglyph) [inline src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java]: Please check the units on the arbitrary feedforward.

File: src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+    @Override
+    public void setVelocity(double velocityMetersPerSecond, double arbitraryFeedForwardVolts) {
+        if (mLastControlMode != ControlType.kVelocity) {
+            mLastControlMode = ControlType.kVelocity;
+        }
+
+        double velocityNative = mSensorModel.toNativeUnits(velocityMetersPerSecond);
+        mSparkMax.getPIDController().setReference(velocityNative, ControlType.kVelocity, kVelocitySlotIdx,
+        arbitraryFeedForwardVolts / mVoltageCompSaturation);
```
Comment (pietroglyph) [inline src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java]: Please check units on arbitrary feedforward.

File: src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    public void setVelocityGains(double kP, double kI, double kD, double kF) {
+        mSparkMax.getPIDController().setP(kP, kVelocitySlotIdx);
+        mSparkMax.getPIDController().setI(kI, kVelocitySlotIdx);
+        mSparkMax.getPIDController().setD(kD, kVelocitySlotIdx);
+        mSparkMax.getPIDController().setFF(kF, kVelocitySlotIdx);
+    }
+
+    @Override
+    public void setPosition(double positionMeters) {
+        if (mLastControlMode != ControlType.kPosition) {
```
Comment (pietroglyph) [inline src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java]: You don't need this if statement.

File: src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+    @Override
+    public void setUseMotionProfileForPosition(boolean useMotionProfile) {
+        mUseMotionProfileForPosition = useMotionProfile;
+    }
+
+    @Override
+    public void setDutyCycle(double dutyCycle, double arbitraryFeedForwardVolts) {
+        if (mLastControlMode != ControlType.kDutyCycle) {
```
Comment (pietroglyph) [inline src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java]: You don't need this if statement.

**Review by pietroglyph — CHANGES_REQUESTED** [review]
Looks mostly good. There are a few unit issues. We need to deal with formatting too, but this is an issue on my side (I have to add a formatter.)

File: src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java (hunk 6 above)
Comment (thehappycamden) [reply src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java]: it was in SpartronicsSRX

File: src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java (hunk 5 above)
Comment (thehappycamden) [reply src/main/java/com/spartronics4915/lib/hardware/motors/SpartronicsMax.java]: it was in SpartronicsSRX

**Pull request merged**